File: src/android/io/github/pwlin/cordova/plugins/pdialog/PDialog.java
/*
The MIT License (MIT)

Copyright (c) 2014 pwlin - pwlin05@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package io.github.pwlin.cordova.plugins.pdialog;

import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONArray;

import android.app.ProgressDialog;

// import android.util.Log;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

public class PDialog extends CordovaPlugin {

	// private static String LOG_TAG = "CordovaLog";
	private static ProgressDialog pDialogObj = null;

	/**
	 * Executes the request and returns PluginResult.
	 *
	 * @param action The action to execute.
	 * @param rawArgs The exec() arguments in JSON form.
	 * @param callbackContext The callback context used when calling back into JavaScript.
	 * @return Whether the action was valid.
	 */
	@Override
	public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
		/*
		 * Don't run any of these if the current activity is finishing in order
		 * to avoid android.view.WindowManager$BadTokenException crashing the
		 * app.
		 */
		
		if (this.cordova.getActivity().isFinishing()) {
			return true;
		}
		if (action.equals("init")) {
			this.init(args.getJSONObject(0));
		} else if (action.equals("dismiss")) {
			this.dismiss();
		} else if (action.equals("setProgress")) {
			this.setProgress(args.getInt(0));
		} else if (action.equals("setTitle")) {
			this.setTitle(args.getString(0));
		} else if (action.equals("setMessage")) {
			this.setMessage(args.getString(0));
		} else if (action.equals("setCancelable")) {
			this.setCancelable(args.getBoolean(0));
		}
		return true;
	}

	/**
	 * Initializing the progress dialog and set various parameters
	 * 
	 * @param argsObj
	 * @see https://github.com/pwlin/cordova-plugin-pdialog/blob/master/README.md
	 */
	private void init(final JSONObject argsObj ) {
		final CordovaInterface cordova = this.cordova;
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				if (PDialog.pDialogObj != null) {
					PDialog.pDialogObj.dismiss();
					PDialog.pDialogObj = null;
				}
				
				int theme = 5; // ProgressDialog.THEME_DEVICE_DEFAULT_LIGHT
				if (argsObj.has("theme")) {
					String themeArg = null;
					try {
						themeArg = argsObj.getString("theme");
					} catch (JSONException e) {
						// e.printStackTrace();
					}
					if ("TRADITIONAL".equals(themeArg)) {
						theme = 1; // ProgressDialog.THEME_TRADITIONAL
					} else if ("DEVICE_DARK".equals(themeArg)) {
						theme = 4; // ProgressDialog.THEME_DEVICE_DEFAULT_DARK
					}
					if ("HOLO_DARK".equals(themeArg)) {
						theme = 2; // ProgressDialog.THEME_HOLO_DARK
					}
					if ("HOLO_LIGHT".equals(themeArg)) {
						theme = 3; // ProgressDialog.THEME_HOLO_LIGHT
					}
				}

				int style = ProgressDialog.STYLE_SPINNER;
				if (argsObj.has("progressStyle")) {
					try {
						if ("HORIZONTAL".equals(argsObj.getString("progressStyle"))) {
							style = ProgressDialog.STYLE_HORIZONTAL;
						}
					} catch (JSONException e) {
						// e.printStackTrace();
					}
				}

				boolean cancelable = true;
				if (argsObj.has("cancelable")) {
					try {
						if (argsObj.getBoolean("cancelable") == false) {
							cancelable = false;
						}
					} catch (JSONException e) {
						// e.printStackTrace();
					}
				}

				String title = "";
				if (argsObj.has("title")) {
					try {
						title = argsObj.getString("title");
					} catch (JSONException e) {
						// e.printStackTrace();
					}
				}

				String message = "";
				if (argsObj.has("message")) {
					try {
						message = argsObj.getString("message");
					} catch (JSONException e) {
						// e.printStackTrace();
					}
				}

				PDialog.pDialogObj = new ProgressDialog(cordova.getActivity(), theme);
				PDialog.pDialogObj.setProgressStyle(style);
				PDialog.pDialogObj.setCancelable(cancelable);
				PDialog.pDialogObj.setCanceledOnTouchOutside(cancelable);
				PDialog.pDialogObj.setTitle(title);
				PDialog.pDialogObj.setMessage(message);
				PDialog.pDialogObj.show();
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Dismiss the progress dialog
	 */
	private void dismiss() {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.dismiss();
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Set the value of the progress bar when progress style is "HORIZONTAL"
	 * 
	 * @param value
	 */
	private void setProgress(final int value) {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.setProgress(value);
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Set the title of the progress dialog
	 * 
	 * @param title
	 */
	private void setTitle(final String title) {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.setTitle(title.replaceAll("^\"|\"$", ""));
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Set the message of the progress dialog
	 * 
	 * @param message
	 */
	private void setMessage(final String message) {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.setMessage(message.replaceAll("^\"|\"$", ""));
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Set the progress max of the progress dialog
	 * 
	 * @param max
	 */
	private void setMax(final String max) {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.setMax(Integer.parseInt(max));
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

	/**
	 * Set whether the progress dialog is calncelable or not
	 * 
	 * @param flag
	 */
	private void setCancelable(final boolean flag) {
		Runnable runnable = new Runnable() {
			@Override
			public void run() {
				PDialog.pDialogObj.setCancelable(flag);
				PDialog.pDialogObj.setCanceledOnTouchOutside(flag);
			};
		};
		this.cordova.getActivity().runOnUiThread(runnable);
	}

}

File: src/ios/PDialog.m
/*
 The MIT License (MIT)
 
 Copyright (c) 2016 Orlando Leite - orlleite gmail
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#import <QuartzCore/QuartzCore.h>

#import "PDialog.h"

#define DIALOG_WIDTH 180.0f
#define DIALOG_PADDING 8.0f
#define DIALOG_HEIGHT 90.0f
#define ANIMATION_TIME 0.3f

@implementation PDialog

- (void) init:(CDVInvokedUrlCommand*)command
{
	bool animated = YES;
	
	if( dialogContainer )
	{
		[dialogContainer removeFromSuperview];
		animated = NO;
	}
	
	if( oldDialogContainer )
	{
		[oldDialogContainer removeFromSuperview];
		oldDialogContainer = nil;
	}
	
	NSDictionary *args = [command.arguments objectAtIndex:0];
	
	UIView *root = self.webView.superview;
	
	dialogContainer = [[UIView alloc] initWithFrame:root.bounds];
	[root addSubview:dialogContainer];
	
	[dialogContainer addSubview:[self background:root.bounds]];
	
	if( [args[@"progressStyle"] isEqualToString:@"SPINNER"] )
	{
		[dialogContainer addSubview:[self spinnerDialog:root.bounds title:args[@"title"] message:args[@"message"]]];
	}
	else if( [args[@"progressStyle"] isEqualToString:@"HORIZONTAL"] )
	{
		[dialogContainer addSubview:[self horizontalDialog:root.bounds title:args[@"title"] message:args[@"message"]]];
	}
	
	if( animated )
	{
		dialogContainer.alpha = 0.0f;
		[UIView animateWithDuration:ANIMATION_TIME animations:^
		 {
			 dialogContainer.alpha = 1.0f;
		 }];
	}
	
	if( !singleFingerTap )
	{
		singleFingerTap = [[UITapGestureRecognizer alloc] initWithTarget:self
																  action:@selector(handleSingleTap:)];
	}
	
	[root addGestureRecognizer:singleFingerTap];
	
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)handleSingleTap:(UITapGestureRecognizer *)recognizer
{
	if( isCancellable ) [self dismiss];
}

- (UIView *) background:(CGRect) rect
{
	UIView *background = [[UIView alloc] initWithFrame:rect];
	background.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin
	| UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
	
	background.backgroundColor = [UIColor blackColor];
	background.alpha = 0.5f;
	
	return background;
}

- (UIView *) spinnerDialog:(CGRect) rect title:(NSString *) title message:(NSString *) message
{
	UIView *dialog = [[UIView alloc] initWithFrame:CGRectMake(
															  (rect.size.width - DIALOG_WIDTH) * 0.5f,
															  (rect.size.height - DIALOG_HEIGHT) * 0.5f,
															  DIALOG_WIDTH, DIALOG_HEIGHT)];
	
	dialog.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f alpha:0.8f];
	dialog.layer.cornerRadius = 5.0f;
	dialog.layer.masksToBounds = YES;
	
	CGFloat height = 40.0f;
	UIActivityIndicatorView *actIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
	actIndicator.center = CGPointMake( DIALOG_WIDTH * 0.5f, 30.0f );
	actIndicator.alpha = 1.0f;
	[actIndicator startAnimating];
	//	initWithFrame:CGRectMake( ( DIALOG_WIDTH - 20.0f ) * 0.5f, 20.0f, 20.0f, 20.0f)];
	[dialog addSubview:actIndicator];
	
	height += 10.0f;
	
	UILabel *label = [self createLabel:title size:15.0f y:height];
	titleLabel = label;
	[dialog addSubview:label];
	height += label.frame.size.height + DIALOG_PADDING * 0.5f;
	
	label = [self createLabel:message size:13.0f y:height];
	messageLabel = label;
	[dialog addSubview:label];
	height += label.frame.size.height + DIALOG_PADDING + 10.0f;
	
	CGRect frame = dialog.frame;
	frame.size.height = height;
	frame.origin.y = (rect.size.height - height) * 0.5f;
	dialog.frame = frame;
	
	return dialog;
}

- (UIView *) horizontalDialog:(CGRect)rect title:(NSString *)title message:(NSString *)message
{
	UIView *dialog = [[UIView alloc] initWithFrame:CGRectMake(
															  (rect.size.width - DIALOG_WIDTH) * 0.5f,
															  (rect.size.height - DIALOG_HEIGHT) * 0.5f,
															  DIALOG_WIDTH, DIALOG_HEIGHT)];
	
	dialog.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f alpha:0.8f];
	dialog.layer.cornerRadius = 5.0f;
	dialog.layer.masksToBounds = YES;
	
	CGFloat height = 10.0f;
	
	UILabel *label = [self createLabel:title size:15.0f y:height];
	titleLabel = label;
	[dialog addSubview:label];
	height += label.frame.size.height + DIALOG_PADDING * 0.5f;
	
	label = [self createLabel:message size:13.0f y:height];
	messageLabel = label;
	[dialog addSubview:label];
	height += label.frame.size.height + DIALOG_PADDING;
	
	progressBar = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleDefault];
	[progressBar setFrame:CGRectMake( DIALOG_PADDING, height, DIALOG_WIDTH - DIALOG_PADDING * 2.0f, 2.0f)];
	[progressBar setProgress:0.0f];
	progressBar.trackTintColor = [UIColor grayColor];
	progressBar.progressTintColor = [UIColor whiteColor];
	[dialog addSubview:progressBar];
	
	height += 10.0f + DIALOG_PADDING;
	
	CGRect frame = dialog.frame;
	frame.size.height = height;
	frame.origin.y = (rect.size.height - height) * 0.5f;
	dialog.frame = frame;
	
	return dialog;
}

- (UILabel *) createLabel:(NSString *) text size:(CGFloat)fontSize y:(CGFloat)y
{
	UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(DIALOG_PADDING, y, DIALOG_WIDTH - DIALOG_PADDING * 2.0f, 10.0f)];
	label.adjustsFontSizeToFitWidth = NO;
	label.numberOfLines = 0;
	label.textAlignment = NSTextAlignmentCenter;
	label.backgroundColor = [UIColor clearColor];
	label.textColor = [UIColor whiteColor];
	
	label.font = [UIFont systemFontOfSize:fontSize];
	label.text = text;
	
	[label sizeToFit];
	
	CGRect frame = label.frame;
	frame.size.width = DIALOG_WIDTH - DIALOG_PADDING * 2.0f;
	label.frame = frame;
	
	return label;
}

- (void) dismiss:(CDVInvokedUrlCommand*)command
{
	[self dismiss];
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) dismiss
{
	if( dialogContainer )
	{
		oldDialogContainer = dialogContainer;
		dialogContainer = titleLabel = messageLabel = progressBar = nil;
		
		[UIView animateWithDuration:ANIMATION_TIME
						 animations:^
		 {
			 oldDialogContainer.alpha = 0.0f;
		 }
						 completion:^(BOOL finished)
		 {
			 if( oldDialogContainer )
				 [oldDialogContainer removeFromSuperview];
		 }];
	}
}

- (void) setProgress:(CDVInvokedUrlCommand*)command
{
	if( progressBar )
	{
		CGFloat value = ((CGFloat) [[command.arguments objectAtIndex:0] integerValue])  / 100.0f;
		[progressBar setProgress:value animated:YES];
	}
	
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) setTitle:(CDVInvokedUrlCommand*)command
{
	if( titleLabel )
	{
		titleLabel.text = [command.arguments objectAtIndex:0];
	}
	
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) setMessage:(CDVInvokedUrlCommand*)command
{
	if( messageLabel )
	{
		messageLabel.text = [command.arguments objectAtIndex:0];
	}
	
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) setMax:(CDVInvokedUrlCommand*)command
{
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) setCancelable:(CDVInvokedUrlCommand*)command
{
	isCancellable = [[command.arguments objectAtIndex:0] boolValue];
	
	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@""];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end
